packages/opentelemetry-exporter-collector/src/transform.ts

```diff
@@ -0,0 +1,241 @@
+  name: string
+): collectorTypes.TruncatableString {
+  const value = name.substr(0, OT_MAX_STRING_LENGTH);
+  const truncatedByteCount =
+    name.length > OT_MAX_STRING_LENGTH ? name.length - OT_MAX_STRING_LENGTH : 0;
+
+  return { value, truncatedByteCount };
+}
+
+/**
+ * convert attributes
+ * @param attributes
+ * @param maxAttributes - default value is 30
+ */
+export function toCollectorAttributes(
+  attributes: Attributes,
+  maxAttributes: number = OT_MAX_ATTRIBUTES
+): collectorTypes.Attributes {
+  const attributeMap: collectorTypes.AttributeMap = {};
+  let droppedAttributesCount = 0;
+
+  const keys = Object.keys(attributes || {});
+
+  const countKeys = Math.min(keys.length, maxAttributes);
+  if (keys.length > maxAttributes) {
+    droppedAttributesCount = keys.length - maxAttributes;
+  }
+
+  for (let i = 0; i <= countKeys - 1; i++) {
+    const key = keys[i];
+    const eventAttributeValue = toCollectorEventValue(
+      attributes && attributes[key]
+    );
+    if (eventAttributeValue) {
+      attributeMap[key] = eventAttributeValue;
+    }
+  }
+
+  return {
+    droppedAttributesCount,
+    attributeMap,
+  };
+}
+
+/**
+ * convert event value
+ * @param value event value
+ */
+export function toCollectorEventValue(
+  value: unknown
+): collectorTypes.AttributeValue | undefined {
+  const attributeValue: collectorTypes.AttributeValue = {};
+
+  if (typeof value === 'string') {
+    attributeValue.stringValue = toCollectorTruncatableString(value);
+  } else if (typeof value === 'boolean') {
+    attributeValue.boolValue = value;
+  } else if (typeof value === 'number') {
+    if (Math.floor(value) === value) {
+      attributeValue.intValue = value;
+    } else {
+      attributeValue.doubleValue = value;
+    }
+  }
+
+  return attributeValue;
+}
+
+/**
+ * convert events
+ * @param events array of events
+ * @param maxAttributes - maximum number of event attributes to be converted
+ */
+export function toCollectorEvents(
+  events: TimedEvent[],
+  maxAttributes: number = OT_MAX_ATTRIBUTES
+): collectorTypes.TimeEvents {
+  let droppedAnnotationsCount = 0;
+  let droppedMessageEventsCount = 0; // not counting yet as messageEvent is not implemented
+
+  const timeEvent: collectorTypes.TimeEvent[] = events.map(
+    (event: TimedEvent) => {
+      let attributes: collectorTypes.Attributes | undefined;
+
+      if (event && event.attributes) {
+        attributes = toCollectorAttributes(event.attributes, maxAttributes);
+        droppedAnnotationsCount += attributes.droppedAttributesCount || 0;
+      }
+
+      let annotation: collectorTypes.Annotation = {};
+      if (event.name || attributes) {
+        annotation = {};
+      }
+
+      if (event.name) {
+        annotation.description = toCollectorTruncatableString(event.name);
+      }
+
+      if (typeof attributes !== 'undefined') {
+        annotation.attributes = attributes;
+      }
+
+      // const messageEvent: MessageEvent;
```

Comment:
Should we open a tracking issue for us to add message events? @mayurkale22 / @bg451 is that in the spec?

Reply:
AFAIK we can use `event.attributes` to assign the MessageEventType (`RECEIVED` or `SEND`) and `uncompressedSize` and `compressedSize` bytes instead of adding dedicated api for message event. Maybe you should open an issue to define semantic conversions for all of these attributes.